lib/internal/vm/Module.js

```diff
@@ -0,0 +1,145 @@
+'use strict';
+
+const { isContext } = process.binding('contextify');
+const assert = require('assert');
+const errors = require('internal/errors');
+const Loader = require('internal/loader/Loader');
+const { ModuleWrap } = internalBinding('module_wrap');
+
+const {
+  kUninstantiated,
+  kInstantiating,
+  kInstantiated,
+  kEvaluating,
+  kEvaluated,
+  kErrored,
+} = internalBinding('module_wrap');
+
+let globalModuleId = 0;
+const kWrap = Symbol('wrap');
+const kContextModuleCount = Symbol('contextModuleCount');
+const kLoaderContextInjection = '____loaderContextInjection____';
+
+class Module {
+  constructor(src, options = {}) {
+    let context;
+    if (options.context !== undefined && isContext(options.context.sandbox)) {
+      context = options.context.sandbox;
+      if (options.context[kContextModuleCount] === undefined)
+        throw new errors.TypeError('ERR_CONTEXT_NOT_MODULE_CONTEXT');
+    }
+
+    const url = options.url ||
+      `vm:module(${context !== undefined ?
+        options.context[kContextModuleCount]++ :
+        globalModuleId++})`;
+
+    const wrap = new ModuleWrap(src, url, {
+      context,
+      lineOffset: options.lineOffset,
+      columnOffset: options.columnOffset,
+    });
+
+    Object.defineProperties(this, {
+      [kWrap]: { value: wrap },
+      context: { value: options.context },
+    });
+  }
+
+  get status() {
+    return this[kWrap].getStatus();
+  }
+
+  get url() {
+    return this[kWrap].url;
+  }
+
+  get namespace() {
+    assert(this[kWrap].getStatus() === kEvaluated, 'Module has not evaluated');
+    return this[kWrap].namespace();
+  }
+
+  async link(fn) {
+    const promises = [];
+    this[kWrap].link(async (specifier) => {
+      const p = fn(specifier);
+      promises.push(p);
+      const m = await p;
+      return m[kWrap];
+    });
+    return Promise.all(promises).then(() => undefined);
+  }
+
+  async instantiate() {
+    this[kWrap].instantiate();
```

Comment:
This needs to ensure the current status is `uninitialized`, and linking has fully completed.

Reply:
(Otherwise the C++ code will blow up)